xds/internal/client/v2client_rds.go

```diff
@@ -94,42 +95,154 @@ func generateRDSUpdateFromRouteConfiguration(rc *xdspb.RouteConfiguration, host
+		match := dr.GetMatch()
+		if match == nil {
+			return rdsUpdate{}, fmt.Errorf("matched virtual host's default route doesn't have a match")
+		}
+		if prefix := match.GetPrefix(); prefix != "" && prefix != "/" {
+			// The matched virtual host is invalid. Match is not "" or "/".
+			return rdsUpdate{}, fmt.Errorf("matched virtual host's default route is %v, want Prefix empty string or /", match)
+		}
+		if caseSensitive := match.GetCaseSensitive(); caseSensitive != nil && !caseSensitive.Value {
+			// The case sensitive is set to false. Not set or set to true are both
+			// valid.
+			return rdsUpdate{}, fmt.Errorf("matched virtual host's default route set case-sensitive to false")
+		}
+		route := dr.GetRoute()
+		if route == nil {
+			return rdsUpdate{}, fmt.Errorf("matched route is nil")
+		}
+
+		if wc := route.GetWeightedClusters(); wc != nil {
+			m, err := weightedClustersProtoToMap(wc)
+			if err != nil {
+				return rdsUpdate{}, fmt.Errorf("matched weighted cluster is invalid: %v", err)
+			}
+			return rdsUpdate{weightedCluster: m}, nil
+		}
+
+		// When there's just one cluster, we set weightedCluster to map with one
+		// entry. This mean we will build a weighted_target balancer even if there's
+		// just one cluster.
+		//
+		// Otherwise, we will need to switch the top policy between weighted_target
+		// and CDS. In case when the action changes between one cluster and multiple
+		// clusters, changing top level policy means recreating TCP connection every
+		// time.
+		return rdsUpdate{weightedCluster: map[string]uint32{route.GetCluster(): 1}}, nil
 	}
-	route := dr.GetRoute()
-	if route == nil {
-		return rdsUpdate{}, fmt.Errorf("matched route is nil")
+
+	routes, err := routesProtoToSlice(vh.Routes)
+	if err != nil {
+		return rdsUpdate{}, fmt.Errorf("received route is invalid: %v", err)
 	}
+	return rdsUpdate{routes: routes}, nil
+}
 
-	if wc := route.GetWeightedClusters(); wc != nil {
-		m, err := weightedClustersProtoToMap(wc)
-		if err != nil {
-			return rdsUpdate{}, fmt.Errorf("matched weighted cluster is invalid: %v", err)
+func routesProtoToSlice(routes []*routepb.Route) ([]*Route, error) {
+	var routesRet []*Route
+
+	for _, route := range routes {
+		match := route.GetMatch()
+		if match == nil {
+			return nil, fmt.Errorf("route %v doesn't have a match", route)
 		}
-		return rdsUpdate{weightedCluster: m}, nil
-	}
 
-	// When there's just one cluster, we set weightedCluster to map with one
-	// entry. This mean we will build a weighted_target balancer even if there's
-	// just one cluster.
-	//
-	// Otherwise, we will need to switch the top policy between weighted_target
-	// and CDS. In case when the action changes between one cluster and multiple
-	// clusters, changing top level policy means recreating TCP connection every
-	// time.
-	return rdsUpdate{weightedCluster: map[string]uint32{route.GetCluster(): 1}}, nil
+		if len(match.GetQueryParameters()) != 0 {
+			// Ignore route with query parameters.
```

Comment:
Should we log a warning? Is this a common event?

Reply:
Done